tox.ini

```diff
@@ -9,7 +9,23 @@ minversion = 1.7.0
 
 [testenv]
 deps =
-    -rtests/requirements.txt
+    attrs
+    pexpect >= 4.8.0
+    pyftpdlib >= 1.5.8
+    pytest
+    pytest-cov==4.0.0
+    pytest-xdist
+    sybil >= 1.3.0  # https://github.com/cjw296/sybil/issues/20#issuecomment-605433422
+    testfixtures
+    uvloop; platform_system != "Windows"
+
+    bpython  # optional for shell wrapper tests
+    brotli; implementation_name != 'pypy'  # optional for HTTP compress downloader middleware tests
+    # 1.1.0 is broken on PyPy: https://github.com/google/brotli/issues/1072
+    brotli==1.0.9; implementation_name == 'pypy'  # optional for HTTP compress downloader middleware tests
+    zstandard; implementation_name != 'pypy'  # optional for HTTP compress downloader middleware tests
+    ipython
+    pywin32; sys_platform == "win32"
```

Comment:
These need to be moved to extra-deps: `brotlicffi` (added after you created your PR) , `zstandard`, `uvloop`, `bpython`, `ipython`.

These are required by some non-skippable tests and need to stay here: `pytest` and its plugins, `sybil`, [`pexpect`](https://github.com/scrapy/scrapy/commit/d19e315b0b07b070f45cf3c082d4dfe0e6a87186), `mitmproxy`, `testfixtures`, [`pyftpdlib`](https://github.com/scrapy/scrapy/commit/e70975f0bb4e8378bf0e00ae7a7014247cd59441), [`pywin32`](https://github.com/scrapy/scrapy/commit/19ad94105f70a26ce47683d488b66fb77b79067b).

Reply:
Doing it!

Reply:
I made pyftpdlib optional temporarily when it wasn't ready for 3.12, we can reintroduce that code if we want. mitmproxy tests are also skippable I think?

Reply:
As an aside, I wondered before if we can get rid of `testfixtures` and use pytest's `caplog` instead, but I'm not sure how to use it as I think we don't have fixture injections in our `unittest` tests.

Reply:
Hey @Gallaecio, I was making some tests and I see that:

Removing `bpython` from the [testenv]deps, causes fail on the following tests:

`FAILED tests/test_commands.py::RunSpiderCommandTest::test_custom_asyncio_loop_enabled_true
FAILED tests/test_crawler.py::CrawlerProcessSubprocess::test_asyncio_enabled_reactor_different_loop
FAILED tests/test_crawler.py::CrawlerProcessSubprocess::test_asyncio_enabled_reactor_same_loop
FAILED tests/test_crawler.py::CrawlerProcessSubprocess::test_custom_loop_asyncio
FAILED tests/test_crawler.py::CrawlerProcessSubprocess::test_custom_loop_asyncio_deferred_signal`

and these:

`FAILED tests/test_utils_display.py::TestDisplay::test_pformat - AssertionError: "{'a': 1}" not found in {"{\x1b[33m'\x1b[39;49;00m\x1b[33ma\x1b[39;49;00m\x1b[33m'\x1b[39;49;00m: \x1b[34m1\x1b[39;49;00m}\x1b[37m\x1b[39;49;00m\n",...
FAILED tests/test_utils_display.py::TestDisplay::test_pformat_old_windows - AssertionError: "{'a': 1}" not found in {"{\x1b[33m'\x1b[39;49;00m\x1b[33ma\x1b[39;49;00m\x1b[33m'\x1b[39;49;00m: \x1b[34m1\x1b[39;49;00m}\x1b[37m\x1b[39;49;00m\n",...
FAILED tests/test_utils_display.py::TestDisplay::test_pformat_windows - AssertionError: "{'a': 1}" not found in {"{\x1b[33m'\x1b[39;49;00m\x1b[33ma\x1b[39;49;00m\x1b[33m'\x1b[39;49;00m: \x1b[34m1\x1b[39;49;00m}\x1b[37m\x1b[39;49;00m\n",...
`

Can we keep bpython on the [testenv]deps?

Reply:
Also, I still have to change the versions of the [extra-deps-pinned]deps to the lowest possible version, I was having other troubles because of this problems with bpython and uvloop, I am resuming finding this versions.

Reply:
I have already pinned the lowest version for the libs:

zstandard==0.1
ipython==2.0.0
brotli==0.5.2
brotlicffi==0.8.0

Reply:
Regarding tests not passing when `uvloop` or `bpython` are missing, we should mark those tests to be skipped when those packages are not installed.

For `uploop`, that means editing [this](https://github.com/scrapy/scrapy/blob/642af407049a5ce8e76b7999c68333670c8c8622/conftest.py#L84-L95) and replacing its implementation with something like:

```
@pytest.fixture(autouse=True)
def requires_uvloop(request):
    if not request.node.get_closest_marker("requires_uvloop"):
        return
    try:
        import uvloop
    except ImportError:
        pytest.skip("uvloop is not installed")
```

For `bpython`, you can see [here](https://github.com/scrapy/scrapy/blob/642af407049a5ce8e76b7999c68333670c8c8622/tests/test_utils_console.py#L30) how to mark a test to be skipped when it is missing.

Reply:
Take a look please

Reply:
Looks good!

Reply:
> I made pyftpdlib optional temporarily when it wasn't ready for 3.12, we can reintroduce that code if we want. mitmproxy tests are also skippable I think?

@wRAR I think it is OK to treat these 2 the same as pytest, i.e. required for tests.